gdal/port/cpl_userfaultfd.cpp

```diff
@@ -0,0 +1,214 @@
+  char filename[0x1000];
+  uint64_t page_size;
+  void * page_ptr;
+  uint64_t vma_size;
+  void * vma_ptr;
+};
+
+
+void uffd_cleanup(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+
+  if (!ctx) return;
+
+  if (ctx->uffd != -1) {
+    ioctl(ctx->uffd, UFFDIO_UNREGISTER, &ctx->uffdio_register);
+    close(ctx->uffd);
+  }
+  if (ctx->page_ptr && ctx->page_size)
+    munmap(ctx->page_ptr, ctx->page_size);
+  if (ctx->vma_ptr && ctx->vma_size)
+    munmap(ctx->vma_ptr, ctx->vma_size);
+  free(ctx);
+
+  return;
+}
+
+void * fault_handler(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+  struct uffd_msg msg;
+  struct uffdio_copy uffdio_copy;
+  struct pollfd pollfd;
+
+  pthread_cleanup_push(uffd_cleanup, ctx);
+
+  pollfd.fd = ctx->uffd;
+  pollfd.events = POLLIN;
+
+  // Open asset for reading
+  VSILFILE * file = VSIFOpenL(ctx->filename, "rb");
+
+  if (!file) return nullptr;
+
+  // Loop indefinitely
+  while(1) {
+    uintptr_t fault_addr;
+    uint64_t offset;
+
+    // Poll for event
+    if (poll(&pollfd, 1, 16) == -1) break;
+    if ((pollfd.revents & POLLERR) || (pollfd.revents & POLLNVAL)) break;
+    if (!(pollfd.revents & POLLIN)) continue;
+
+    // Read page fault event
+    if (read(ctx->uffd, &msg, sizeof(msg)) < 1) {
+      if (errno == EWOULDBLOCK) continue;
+      else break;
+    }
+
+    fault_addr = msg.arg.pagefault.address & ~(ctx->page_size-1);
+    offset = reinterpret_cast<uint64_t>(fault_addr) - reinterpret_cast<uint64_t>(ctx->vma_ptr);
+
+    // Copy data into page
+    VSIFSeekL(file, offset, SEEK_SET);
+    VSIFReadL(ctx->page_ptr, ctx->page_size, 1, file);
+
+    uffdio_copy.src = reinterpret_cast<uintptr_t>(ctx->page_ptr);
+    uffdio_copy.dst = fault_addr;
+    uffdio_copy.len = static_cast<uintptr_t>(ctx->page_size);
+    uffdio_copy.mode = 0;
+    uffdio_copy.copy = 0;
+    if (ioctl(ctx->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+  }
+
+  // Return resources
+  VSIFCloseL(file);
+
+  pthread_cleanup_pop(1);
+  return nullptr;
+}
+
+/*
+ * Returns nullptr on failure, a valid pointer on success.
+ */
+void * hookup(const char * pszFilename, void ** ppVma, uint64_t * pnVmaSize)
+{
+  VSIStatBufL statbuf;
+  pthread_t thread;
+  struct cpl_uffd_context * ctx = nullptr;
+  struct uffdio_api uffdio_api;
+
+  // Get the size of the asset
+  if (VSIStatL(pszFilename, &statbuf)) return nullptr;
+
+  // Setup the `cpl_uffd_context` struct
+  ctx = static_cast<struct cpl_uffd_context *>(malloc(sizeof(struct cpl_uffd_context)));
+  strncpy(ctx->filename, pszFilename, sizeof((reinterpret_cast<struct cpl_uffd_context *>(0))->filename));
+  ctx->page_size = sysconf(_SC_PAGESIZE);
+  ctx->vma_size = ((statbuf.st_size/ctx->page_size)+1) * ctx->page_size;
+
+  // If the mmap failed, free resources and return
+  ctx->vma_ptr = mmap(nullptr, ctx->vma_size, PROT_READ, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
```

Comment:
there should be a check that vma_size doesn't excess size_t, to detect files > 4 GB on 32 bit builds

Reply:
I added a check to make sure that `ctx->vma_size` has not overflowed and become smaller than `statbuf.st_size`.  I think that if `statbuf.st_size` (or `ctx->vma_size`) is too for the architecture, the `mmap` below will fail.